### ISSUE TITLE
Pojmenované Input Mappingy

### ISSUE THREAD

**Issue by ondrejhlavacek** [keboola/kbc-ui#674]
U některých komponent a OpenRefine transformace by bylo lepší mít fixní předdefinovaný input/output mappingy. Říkám tomu zatím pojmenovaný input mappingy (nic lepšího mě nenapadlo) a jde jen o to, že je zafixovaný `destination` u input mappingů a `source` u output mappingů a uživatel tak vybírá pouze tabulky ve storage.

V connection v konfiguraci komponenty zavádím nové pole `UI Options`, ve kterém by se to definovalo. Komponenta v tomto příkladě na vstupu očekává `in/tables/data.csv` a produkuje `out/tables/data.csv`. 

```json
{
  "input": [
    {
      "label": "Load data from table",
      "destination": "data.csv"
    }
  ],
  "output": [
    {
      "label": "Store data to table",
      "source": "data.csv"
    }
  ]
}
```

V UI se u nově založené konfigurace objeví vložený input i output mapping (bez zadaných údajů), které budou popsány dle `label` a uživatel tam jen vybere tabulku ve storage (source/destination).

Zatím je to narychlo navržené takto, ale mám pár otázek

1) Možnost rozšířit pojmenovaný input mapping i na jiné pole než source/destination? Typicky primární klíč  a inkrementální loady v output mappingu.

2) Co soubory? Neužilo by se tam potenciálně i něco takového? 

3) Umožnit v komponentě, která má pojmenovaný mapping, přidávat i další mappingy, nebo to bude zafixované?

4) Co výběr bucketů/sloupců?

5) Jak správně strukturovat UI Options? (viz 2, 3, 4). Nebo tam možná přijde i něco dalšího? Mergnout tam JSON Schéma, UI flagy, defaultní nastavení procesorů (https://github.com/keboola/docker-bundle/issues/81)?

**Comment by odinuv** [comment]
1) dal by se primarni klic zdedit ze storage?, inkrementalni load by podle me vzdycky mel specifikovany uzivatelem, ale mozna by komponenta mohla davat default?
2) nenapada me zadny use case, ostatne ted se soubory pojmenovavaji automaticky a nikomu to nevadi
3) imho ne, bud ta komponenta ma pevny vstupy, nebo variabilni, nicmene melo by byt mozny, aby komponenta mela vic pojmenovanych mappingu 
4) jo! :)
5) jo pridal bych tam i json schema a ui flagy

```
"uioptions": {
    "input": [
        {
            "label": "Input table",
            "description": "this is the main input table",
            "destination": "data.csv",
        },
        {
            "label": "Supplementary table",
            "description": "something else"
            "destination": "additional.csv"
            "columns": [
                [   
                    "name": "Id column",
                    "decription": "foobar",
                    "configPath": "parameters.columns.id" # misto v konfiguraci, kde se objevi nastavene jmeno sloupce
                ],
                [   
                    "name": "Size",
                    "decription": "foobar",
                    "configPath": "parameters.size"
                ]               
            ]
        }       
    ],
    "tableOutput": ...
}
```

**Comment by ondrejhlavacek** [comment]
@odin 

ad 1) to je taková klasika, moc nevím jak UX na to... pokud výstupní tabulka existuje, PK by se mohl přebírat, ale pokud neexistuje, tak bys ho měl mít možnost nastavit, v output mappingu? dokážu si představit i situace, kdy uživatel změni PK v tabulce a extraktor musí vyfailovat, protože to neladí s jeho vlastní představou o PK (nezaručí potom správný data)

**Comment by odinuv** [comment]
hmno, asi je potreba oddelene resit PK na vstupu a PK na vystupu
- PK na vstupu by se mel asi prebirat vzdycky a uzivatel do toho nema asi co mluvit
- PK na vystupu by asi mela specifikovat komponenta a skoro bych rekl, ze povinne, kdo jinej by to mel vedet lip, kde je PK ?

**Comment by ondrejhlavacek** [comment]
@odin můžeme ty PK přesunout do jiné issue? https://github.com/keboola/kbc-ui/issues/681

**Comment by ondrejhlavacek** [comment]
v UI je to hotový, teď jen pořešit, jak to dostat do konfigurace komponenty

**Comment by odinuv** [comment]
No pokud je to teda v nejakym `uioptions` u komponenty, tak to musi docker runner precist a predhodit do `storage.input` konfigurace  ?

**Comment by ondrejhlavacek** [comment]
docker runner? co ten s tím má co společnýho? to bude dělat UI

**Comment by odinuv** [comment]
no tak, kdyz jsi napsal "v UI je to hotový, teď jen pořešit, jak to dostat do konfigurace komponenty" tak jsem to pochopil tak, ze to ma delat docker runner :)

**Comment by ondrejhlavacek** [comment]
nene, myslel jsem tím, že UI komponenty mám hotový, ale ještě si to odnikud nenačítaj, je to zahardcodovaný v UI, takže teď to jen protáhnout o krok dál a nechat tu konfiguraci UI načítat odněkud (ui options).

**Comment by odinuv** [comment]
cc @JakubMatejka tak to je asi otazka na devportal - nejak chytre rozsirit soucasny ui options https://github.com/keboola/kbc-ui/issues/674#issuecomment-251038162

**Comment by ondrejhlavacek** [comment]
@odinuv ty se asi ani rozšiřovat nemusí, nevím, jestli v nich vůbec něco je?

**Comment by odinuv** [comment]
no jsou v nich ty flagy - `genericDockerUI` aspol.

**Comment by ondrejhlavacek** [comment]
Jo, už to vidím, to je pole pro tagy, tak by to mohl bejt nějakej novej field, I/O config? Nicméně bych celý tohle ještě nějak promyslel, jestli to fakt takhle chceme dělat (a jestli není nějaká lepší cesta).

**Comment by odinuv** [comment]
pole pro tagy? ne - mluvim o bodu 5. :)

jsem myslel, ze uz to mas hotovy https://github.com/keboola/kbc-ui/issues/674#issuecomment-252590079

mozna bude lepsi si casem zavolat

**Comment by ondrejhlavacek** [comment]
Ad UI Options - vede u toho multiselect dropdownu link na https://developers.keboola.com/extend/registration/#ui-options, kde je jen výčet `tagů` (špatně jsem to pojmenoval, jsou to položky multiselectu).

![image](https://user-images.githubusercontent.com/497675/35332053-f62bf12a-0109-11e8-8f4b-ef522e1c129e.png)

Pojmenovaný mappingy jsou v UI připravený, použitý u jedný komponenty (OpenRefine transformace), ale je to děsně starý (a za ten rok a půl to nikdy nikdo nechtěl na něco dalšího použít). Jenom bych se syncnul se současným směřováním UI, jestli to ještě dává smysl takhle. V hlavě se mi u toho honí JSON schémata (+ custom typy), config rows, obecně rozšiřitelnost pro vývojáře zvenčí.

**Comment by odinuv** [comment]
Treba u darksky, geocoding, what3words, by se to uplatnilo, pokud to nebudem resit takhle, tak bych to resil pres vytvoreni `genericDockerUI-singleTableInput` - coz neni moc systematicky, ale bylo by to resenim

**Comment by ondrejhlavacek** [comment]
to by taky mohlo bejt řešení! tyhle komponenty neřeší, jakej soubor mají na vstupu?

**Comment by odinuv** [comment]
potrebuji jednu tabulku, nazev je nezajima (ctou z konfigu), kdyby existoval column selector, tak by tvar te vstupni tabulky mohl byt volnejsi a bylo by to cely robustnejsi, ale soucasnej stav je asi taky prijatelnej - tzn. dulezity je zbavit se toho, aby tam 0..n tabulek a uzivatel zadaval na IM destination

**Comment by ondrejhlavacek** [comment]
To by šlo interně vzájemně propojit. Na tom pojmenovaným IM bylo dobrý, že se to dá celý custom otextovat.

**Comment by Halama** [comment]
u toho darsky se to da obejit tak ze vyselektujes sloupce v  tom IM, to zaroven zohlednuje i poradi ve kterem je vyberes. Ten samotny extractor resi jen poradi, nic jineho. Na ovladani to ale asi moc intuitivni neni hezi by bylo nejake mapovani tech sloupcu.
